src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegment.cs

```diff
@@ -81,6 +72,19 @@ public void ResetMemory()
             _next = null;
             _end = 0;
             AvailableMemory = default;
+
+            // Return the memory, use a fast exact type check rather than checking the inheritance hierarchy
+            // or following the interface mapping.
+            if (memoryOwner.GetType() == typeof(byte[]))
```

Comment:
Such check is fast on .NET 5 only :/

Reply:
Slower than `is` cast to interface on others?

Reply:
Yep. A few times slower on  Core 3.1 and Framework 4.8. I haven't checked on Core 5.0.

Reply:
Packaging issue anyway :( will have to .NET 5 it

Reply:
> **update:** Ignore. I compared it to is `byte[]`

Yeah that will call into `IsInstanceOfClass` and pointer chase the inheritance hierarchy